### PR TITLE
Keycloak 3297

### PR DESCRIPTION
**Pull request by ebondu**

### CODE REVIEW

**Comment by stianst** [comment]
This should only have a single commit. There are also loads of changes that have nothing to do with the CORS headers. Please clean-up your PR.

Further we need PR to documentation as well as integration tests added to be able to accept this contribution.

**Comment by ebondu** [comment]
I updated the PR, let me know if it's ok now.

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/AuthenticatedActionsHandler.java [hunk 1]
```diff
@@ -124,6 +125,7 @@ protected boolean corsRequest()  {
             facade.getResponse().setStatus(200);
             facade.getResponse().setHeader(CorsHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, origin);
             facade.getResponse().setHeader(CorsHeaders.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+            facade.getResponse().setHeader(CorsHeaders.ACCESS_CONTROL_EXPOSE_HEADERS, exposeHeaders);
```
Comment (stianst) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/AuthenticatedActionsHandler.java]: Only set header if "deployment.getCorsExposedHeaders() != null"

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java [hunk 2]
```diff
@@ -109,7 +109,7 @@ public KeycloakAuthenticationProcessingFilter(AuthenticationManager authenticati
      *
      */
     public KeycloakAuthenticationProcessingFilter(AuthenticationManager authenticationManager, RequestMatcher
-                requiresAuthenticationRequestMatcher) {
+            requiresAuthenticationRequestMatcher) {
```
Comment (stianst) [inline adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java]: Remove formatting changes

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java [hunk 3]
```diff
@@ -142,7 +142,7 @@ public Authentication attemptAuthentication(HttpServletRequest request, HttpServ
         if (AuthOutcome.FAILED.equals(result)) {
             throw new KeycloakAuthenticationException("Auth outcome: " + result);
         }
-       
+
```
Comment (stianst) [inline adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java]: Remove formatting changes

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java [hunk 4]
```diff
@@ -184,7 +184,7 @@ protected boolean isBasicAuthRequest(HttpServletRequest request) {
 
     @Override
     protected void successfulAuthentication(HttpServletRequest request, HttpServletResponse response, FilterChain chain,
-            Authentication authResult) throws IOException, ServletException {
+                                            Authentication authResult) throws IOException, ServletException {
```
Comment (stianst) [inline adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java]: Remove formatting changes

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java [hunk 5]
```diff
@@ -212,7 +212,7 @@ protected void successfulAuthentication(HttpServletRequest request, HttpServletR
 
     @Override
     protected void unsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response,
-            AuthenticationException failed) throws IOException, ServletException {
+                                              AuthenticationException failed) throws IOException, ServletException {
```
Comment (stianst) [inline adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java]: Remove formatting changes

File: adapters/oidc/spring-security/src/test/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilterTest.java [hunk 6]
```diff
@@ -244,6 +244,6 @@ private void setBasicAuthHeader(MockHttpServletRequest request) {
     }
 
     private void setAuthorizationHeader(MockHttpServletRequest request, String scheme) {
-      request.addHeader(KeycloakAuthenticationProcessingFilter.AUTHORIZATION_HEADER, scheme + " " + UUID.randomUUID().toString());
+        request.addHeader(KeycloakAuthenticationProcessingFilter.AUTHORIZATION_HEADER, scheme + " " + UUID.randomUUID().toString());
```
Comment (stianst) [inline adapters/oidc/spring-security/src/test/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilterTest.java]: Remove formatting changes

**Review by stianst — COMMENTED** [review]
Tests are missing and documentation is not updated. You also need to add the property to the adapter subsystem based configuration.

**Comment by ebondu** [comment]
Regarding the documentation, can I create a separate PR on keycloak/keycloak-documentation repo?
I will change this PR to integrate your comments (formatting changes + the Null test). Do you expect more unit tests ? (I already add them, for me it was ok)
Can you give me more details (classes / adapters to modify) to 'add the property to the adapter subsystem based configuration' ?

**Comment by sebastienblanc** [comment]
@ebondu Could you remove the format diffs and add the property to the subsystem ? Would be nice to have this merged since it adds a nice feature.

**Comment by ebondu** [comment]
I am ready to commit changes regarding the format diffs but I don't understand what you means by adding the property to the subsystem ?

**Comment by sebastienblanc** [comment]
@ebondu You adding a new property to the core adapter, which can also be configured via wildfly subsystem, looks at this PR where I added also a new property : https://github.com/keycloak/keycloak/pull/4035/

**Comment by ebondu** [comment]
Ok thk. Regarding the doc, I prepared a separate PR for the keycloak-documentation repository, is it the proper way ?

**Comment by sebastienblanc** [comment]
yes

File: adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java (hunk 3 above)
Comment (sebastienblanc) [inline adapters/oidc/spring-security/src/main/java/org/keycloak/adapters/springsecurity/filter/KeycloakAuthenticationProcessingFilter.java]: You forgot this one ;)

**Comment by ebondu** [comment]
The corresponding [PR](https://github.com/keycloak/keycloak-documentation/pull/123) for the doc.

**Comment by sebastienblanc** [comment]
Thanks but there still a few issues : 
* https://github.com/keycloak/keycloak/pull/3759/commits/c3feb211c75e32737f51d7914cc694354c8ac9a8 and https://github.com/keycloak/keycloak/tree/f058e0f830ba0702bfe9649bd91e1bd496bd50dd commits (the 2 first commits) should not be in this PR. 
* A rebase with master is needed
* A small formatting change that hasn't been removed

**Comment by sebastienblanc** [comment]
@ebondu  Would be nice if we could get that in for the 3.1 release, if you address the 3 points I mentioned before it should be doable.

**Comment by ebondu** [comment]
I just rebased the branch and all other issues was suposed to be solved since last commit.

**Comment by sebastienblanc** [comment]
This looks all good, thx @ebondu for the contribution
@stianst can you merge ?

**Comment by stianst** [comment]
This is a bit of a mess. There's 9 commits here and seems like there's fixes for 3 separate JIRAs as well.

**Comment by sebastienblanc** [comment]
Hum, right, I thought he had removed the two first commits ... 
@ebondu could you remove these 
* https://github.com/keycloak/keycloak/pull/3759/commits/c3feb211c75e32737f51d7914cc694354c8ac9a8 and https://github.com/keycloak/keycloak/tree/f058e0f830ba0702bfe9649bd91e1bd496bd50dd commits (the 2 first commits) should not be in this PR. 

And yes, if you can squash the rest of the commtis into 1 commit.

**Comment by ebondu** [comment]
Sorry but I am lost ^^. Based on the 14 final modified files, I was thinking the final result was OK.

**Comment by sebastienblanc** [comment]
Yes, that is how I got confused, in my latest review I just reviewed the change files which is indeed only related to this ticket but for some reasons old commits are still hanging there. 
My advise is to create a fresh branch from master and just add your commits related to this change (and so you can squash them nicely into 1 single commit)

**Comment by sebastienblanc** [comment]
Ok let me close this one, I created a new PR based on your commits : https://github.com/keycloak/keycloak/pull/4074

**Comment by ebondu** [comment]
Ok thanks ;)

**Comment by sebastienblanc** [comment]
@stianst looks like I have no right to close PRs , can you close this one ?

**Pull request closed without merging**